src/settings/SettingsStore.js

```diff
@@ -98,6 +100,121 @@ const LEVEL_ORDER = [
+    // relevant to the setting. We pass the watcher on to the handlers and aggregate it
+    // before sending it off to the caller. We need to track which callback functions we
+    // provide to the handlers though so we can unwatch it on demand. In practice, we
+    // return a "callback reference" to the caller which correlates to an entry in this
+    // dictionary for each handler's callback function.
+    //
+    // We also maintain a list of monitors which are special watchers: they cause dispatches
+    // when the setting changes. We track which rooms we're monitoring though to ensure we
+    // don't duplicate updates on the bus.
+    static _watchers = {}; // { callbackRef => { level => callbackFn } }
+    static _monitors = {}; // { settingName => { roomId => callbackRef } }
+
+    /**
+     * Watches for changes in a particular setting. This is done without any local echo
+     * wrapping and fires whenever a change is detected in a setting's value. Watching
+     * is intended to be used in scenarios where the app needs to react to changes made
+     * by other devices. It is otherwise expected that callers will be able to use the
+     * Controller system or track their own changes to settings. Callers should retain
+     * @param {string} settingName The setting name to watch
+     * @param {String} roomId The room ID to watch for changes in. May be null for 'all'.
+     * @param {function} callbackFn A function to be called when a setting change is
+     * detected. Four arguments can be expected: the setting name, the room ID (may be null),
+     * the level the change happened at, and finally the new value for those arguments. The
+     * callback may need to do a call to #getValue() to see if a consequential change has
+     * occurred.
+     * @returns {string} A reference to the watcher that was employed.
+     */
+    static watchSetting(settingName, roomId, callbackFn) {
+        const setting = SETTINGS[settingName];
+        const originalSettingName = settingName;
+        if (!setting) throw new Error(`${settingName} is not a setting`);
+
+        if (setting.invertedSettingName) {
+            settingName = setting.invertedSettingName;
+        }
+
+        const watcherId = `${new Date().getTime()}_${settingName}_${roomId}`;
+        SettingsStore._watchers[watcherId] = {};
+
+        const levels = Object.keys(LEVEL_HANDLERS);
+        for (const level of levels) {
+            const handler = SettingsStore._getHandler(originalSettingName, level);
+            if (!handler) continue;
+
+            const localizedCallback = (changedInRoomId, newVal) => {
+                callbackFn(originalSettingName, changedInRoomId, level, newVal);
+            };
+
+            console.log(`Starting watcher for ${settingName}@${roomId || '<null room>'} at level ${level}`);
+            SettingsStore._watchers[watcherId][level] = localizedCallback;
+            handler.watchSetting(settingName, roomId, localizedCallback);
```

Comment:
Hmm, so if a handler at a lower level emits an update, but a higher level handler would still override that value, the update still goes out, right? I wonder why the handlers each have their own `WatchManager`, as opposed to them having a reference (or use the global) to the `SettingsStore` to call `notifyUpdate` on... where you can then decide to emit an update bases on all the levels.

Not a huge issue, as most code would just cal `getValue` again and read the same old value, but on first sight that would also be less code, would be easier to prevent the above "issue", but might be missing something.

At least I would remove/change the newVal argument to the callback, because I suspect that as it stands it might report something different than what `getValue` would report.

Reply:
The idea is to maintain the high flexibility provided by the SettingsStore for the admittedly few cases which need it. There's not currently a use case for needing to know when a specific level has changed, however things like notifications (which currently use a controller) might want to make use of this granularity. I'm a bit on the fence for changing this to only emit when there's a perceivable change as in future it would result in excessive updates but might be inconsequential.

The other reason each handler has a dedicated WatchManager is to support the idea of watching a setting at a particular level. This is also a rare case, and we can probably just split it out later if we need to.